OKImageDownloader/Classes/API/ImageDownloader.swift

```diff
@@ -69,9 +69,10 @@ public final class ImageDownloader: ImageDownloading {
     
     // MARK: - Downloading
     
-    @discardableResult
-    public func download(url: URL, completionHandler: @escaping CompletionHandler) -> ImageDownloaderReceipt {
+    public func download(url: URL, receiptHandler: ImageDownloaderReceiptHandling? = nil, completionHandler: @escaping CompletionHandler) {
         let receipt = ImageDownloaderReceipt(id: UUID(), url: url)
+        var receiptHandler = receiptHandler
```

Comment:
You should make the `ImageDownloaderReceiptHandling` a class protocol or just pass an `inout` argument that you can mutate. If someone passes a `struct` this will create a copy on write and it won't modify the original `receiptHandler`. As an alternative you could make  `ImageDownloaderReceiptHandling` have a function instead `func handle(receipt: ImageDownloaderReceipt)` to escape the mutability semantics.

Reply:
@juliantejera  A func on a struct would still not fix the COW aspect of the mutability, so the `inout` is the only option I think.

Reply:
So it turns out you can't use `nil` with inout since nil isn't mutable 🙄 So class protocol it is...